source/containerimage/pull.go

```diff
@@ -200,10 +200,7 @@ func (p *puller) CacheKey(ctx context.Context, g session.Group, index int) (cach
 		}
 
 		if len(p.manifest.Descriptors) > 0 {
-			pw, _, _ := progress.FromContext(ctx)
-			progressController := &controller.Controller{
-				Writer: pw,
-			}
+			progressController := &controller.Controller{}
```

Comment:
I think this loses the correct context for the progress. Eg. lazy blob pulls happening later should still be associated with the image where they are coming from. I think we need to read out the progress writer from context in here but not initialize it yet. Only read out constructor/callback.

Reply:
It is confusing with the multiple contexts involved here, I was hoping the `ctx` passed into the `Start` would be the right one, but sounds like not.  If my understanding is correct, I think we have two options:
1) Capture `ctx` at the time we create the `&controller.Controller{}`, perhaps store a `progressCtx` in the struct for later use?  For this I would just add `func NewController(context.Context) progress.Controller`

2) Split `progress.FromContext` into two parts, one to just return the progress stored in `ctx.Value()` so we can stash it in the `Controller`, and then another that does the `newWriter`. If we go this route I think I would rename `progress.FromContext` to `progress.NewFromContext`.  Then add a new `progress.FromContext` that would just load progress from `ctx.Value`.  Then rename `newWriter(Writer)` to `NewWriter(writer)` so it can be used outside the package.

Thoughts?

Reply:
I went with option 2) and updated the PR, if you disagree should be easy enough to change.